src/interface/entities.ts

```diff
@@ -0,0 +1,173 @@
+/*
+ * Copyright (c) 2020 Broadcom.
+ * The term "Broadcom" refers to Broadcom Inc. and/or its subsidiaries.
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *   Broadcom, Inc. - initial API and implementation
+ */
+
+import { IProfile } from '@zowe/imperative';
+import { IEndevorQualifier } from './IEndevorQualifier';
+import { IFilter } from './IFilter';
+
+export interface IConnection extends IProfile {
+  name?: string | undefined;
+  host?: string | undefined;
+  port?: number | undefined;
+  user?: string | undefined;
+  password?: string | undefined;
+  rejectUnauthorized?: boolean | undefined;
+  protocol?: string | undefined;
+}
+
+export interface IElement extends IEndevorEntity {
+  elmName: string;
+  fullElmName: string;
+  elmVVLL: string;
+  envName: string;
+  sysName: string;
+  sbsName: string;
+  stgNum: string;
+  typeName: string;
+  repository: IRepository;
+  getName: () => string;
+  getDescription: () => string;
+  getElmName: () => string;
+  getElmVVLL: () => string;
+  getRepository: () => IRepository;
+  getQualifier: () => IEndevorQualifier;
+}
+
+export interface IEndevorEntity {
+  filters?: IEndevorFilter[];
+  getName(): string | undefined;
+  getDescription(): string | undefined;
+  getRepository(): IRepository;
+}
+
+export interface IEndevorFilter {
+  loadElements: (newElements: IElement[], append: boolean) => void;
+  getName: () => string;
+  getUri: () => string;
+  getDescription: () => string;
+  getRepository: () => IRepository;
+  setRepository: (value: IRepository) => void;
+  getElements: () => IElement[];
+  getQualifier: () => IEndevorQualifier;
+  updateFilterString: (filterString: string) => void;
+  editFilter: (name: string) => void;
+  deleteFilter: () => void;
+}
+
+export interface IEnvironment {
+  envName: string;
+  repository: IRepository;
+  systems: Map<string, ISystem>;
+  stages: IStage[];
+  loadSystems: (newSystems: ISystem[], append: boolean) => void;
+  loadStages: (newStages: IStage[]) => void;
+  getName: () => string;
+  getDescription: () => string;
+  getEnvName: () => string;
+  getRepository: () => IRepository;
+  findSystem: (sysName: string) => ISystem | undefined;
+  getSystems: () => ISystem[];
+  getStage: (num: number) => IStage | undefined;
+  getStages: () => IStage[];
+}
+
+export interface IRepository {
```

Comment:
Url, name, and datasource are all private properties of the repo class, therefore they are not included in the interface. We could make them public in the class, but that should be considered during a refactoring of that class, which is out of the scope of this PR.

Reply:
Yeah, I talked more about getters here. I suppose, the accessor 'getUrl' cannot return undefined and if it return - something went wrong with the entity, because the repository without url doesn't make any sense for me. 

Ok, I agree, that the point is in more global scope...